### DISCUSSION TITLE
Muliple schema migration via custom command

### DISCUSSION THREAD

**Discussion by PrinceLoren** [mikro-orm/mikro-orm#3969] [Q&A] [answered]
Hi, i am using custom cli command to create migration, now I have left only the migration creation case to minimize confusion.
My database is `postgresql`

``` javascript
export class OrmRunner implements CommandRunner {
  private migrator: IMigrator;
  private schemaGenerator: ISchemaGenerator;
  private entityGenerator: IEntityGenerator;

  private validateAction(action: string): asserts action is Action {
    if (!actions.includes(action as Action)) {
      throw new Error(`Unknown action \`${action}\``);
    }
  }

  constructor(private orm: MikroORM<PostgreSqlDriver>) {
    this.migrator = this.orm.getMigrator();
    // this.schemaGenerator = this.orm.getSchemaGenerator()
    // this.entityGenerator = this.orm.getEntityGenerator();
  }

  async run([action]: string[]): Promise<void> {
    this.validateAction(action);

    switch (action) {
      case 'create_migration': {
        await this.migrator.createMigration();
        return;
      }
    }
  }
}
```

i have 5 schemas in database include public, but when i create migration,migration is created only for public like this

``` javascript
export class Migration20230118120100 extends Migration {
  async up(): Promise<void> {
    this.addSql(‘alter table “invoice” add column “is_changed” boolean null;’);
  }
  async down(): Promise<void> {
    this.addSql(‘alter table “invoice” drop column “is_changed”;’);
  }
}
```

as far as I understand, the create Migration() function creates migrations only for the current schema, is it possible to change the function so that it creates migrations for all schemas at once or do it like this

``` javascript
  async up(schema): Promise<void> {
    this.addSql(`alter table ${schema}.“invoice” add column “is_changed” boolean null;`);
  }
```

### it follows from all this that when I create columns in an entity, in the database, they appear only in the public schema

now, when configuring the database, I do not specify the schemas, my current config 

``` javascript
  const database = {
    type: 'postgresql',
    host: process.env.POSTGRES_HOST,
    port: process.env.POSTGRES_PORT,
    user: process.env.POSTGRES_USER,
    password: process.env.POSTGRES_PASSWORD,
    dbName: process.env.POSTGRES_DB,
    pool: {
      max: process.env.POSTGRES_MAX_CONNECTIONS,
    },
  };
```
and then just  

``` javascript
  imports: [
        MikroOrmModule.forRootAsync({
          useFactory: (config: ConfigService) =>
            merge(
              {
                ...config.get('database'),
                entities,
                migrations: {
                  migrationsList,
                },
              }
            ),
          inject: [ConfigService],
          imports: [ConfigModule],
        }),
      ],
``` 

I can change the scheme in any place convenient for me, for example

``` javascript
  case 'create_migration': {
        const schemas = ['public', '1', '2', '3', '4']
        for (const schema of schemas) {
          await this.orm.config.set('schema', schema)
          await this.migrator.createMigration();
        }
        return;
      }
```
I even tried this option, but unfortunately it does not work properly, eventually, the names of these schemes should appear in the namespace in the snapshot

``` json
  "namespaces": ["public", "1", "2", "3", "4"],
  "name": "public",
  "tables":
```

**Comment by B4nan** [comment, marked as answer]
This is a rather unsupported use case, as so far there was nobody able to propose something reasonable - people always ask for one hack here or there, without thinking about the bigger picture. Migrations are static, while your set of existing schema is not (I guess), and that is the reason why you will always need to handle this on your app level, the ORM doesn't know what you expect from your setup. Migrations are maintained in a table, that can either live onces in the main/public schema, or be there for every one of the schemas you maintain, again something the ORM is not aware of. Lastly you might want to have the entity in main schema only, in all schemas, or only in the "child schemas" (so not in the main one). A bit too many variables to have a general solution... Note that the schema generator works with _many schemas_ at the same time, as you can as well have multiple static schemas (not a tenant per client), and the schema generator will read all of them for you.

What does your invoice entity look like, do you use wild card schema on it? That is the only concept we have available for this, it's not perfect, but it could help.

https://mikro-orm.io/docs/next/multiple-schemas#wildcard-schema

**Reply by PrinceLoren** [reply]
At the moment it looks like this

``` javascript
@Entity({ customRepository: () => OfferRepository })
```

But I've also tried doing this

``` javascript
@Entity({ customRepository: () => OfferRepository, schema: '*' })
```

But I read in the documentation that if I do this, it will be ignored during migration

**Reply by PrinceLoren** [reply]
I solved the problem with different schemas when I want to add data to different schemas, but I can't get the new columns to show up in all schemas

**Reply by B4nan** [reply]
Yes, because it does not make much sense to maintain the migration state for such. Again, the ORM dont know about what schemas do you want to maintain, on the other hand it sounds like a sane expectation that you want to keep them all in sync with metadata - you dont need migrations for that, schema generator is enough, and you can use that dynamically.

**Reply by B4nan** [reply]
> but I can't get the new columns to show up in all schemas

So you expect to persist some entity and somehow propagate it to all the schemas? Are you sure you are using the right tool for the job? First time hearing such requirement for a multi tenant app, you are probably designing something wrong.

**Reply by PrinceLoren** [reply]
Let me write down exactly what I'm doing so as not to confuse you

1) I am creating a new field in entity

``` javascript 
@Property({fieldName: 'is_changed', nullable: true })
isChanged?: boolean;
```

2) then I create a migration
3) then when the application is launched, `migrator.up()` is called
4) in the database in the public schema in this table I have this field (is_changed)
5) but if I go to another scheme, then in the same table I do not have this field (is_changed)

**Reply by PrinceLoren** [reply]
It just seemed to me that it's quite expected that if I don't specify in entity which specific schema to use, then the column should appear in all schemas in the database.

**Reply by B4nan** [reply]
> It just seemed to me that it's quite expected that if I don't specify in entity which specific schema to use, then the column should appear in all schemas in the database.

That's not how it works, if you omit the schema name from the decorator, it means it's the default schema - for postgres it is the `public` one, unless you change it in the ORM config. You want wildcard schema for this, which implies the entity living in many schemas, not just once. And just like you are trying to bend the migrations layer for such, you should instead use the schema generator, where you can set the schema dynamically. 

I know its not perfect, but it makes more sense to me than bending migrator, which is designed to work statically, the whole point of migrations is to be able to verify the queries before you run them. But for that it needs to be clear on what schema the migration works on.

Maybe one last point on why the migrator cant handle the wildcard entities - as you have many schemas, they can be out of sync, some can be newly added (hence empty), some can be dated (for whatever reason, e.g. last migration failed), some can be up to date already. You can't have one statically written migration script and expect things to work on all the schemas. That was the thinking I had when designing the feature, to me it makes the most sense to just do `orm.schema.updateSchema({ schema: '...' })` to always make them in sync with the metadata, regardless of their current state. To have some kind of control over this, you could still use first the `getUpdateSchemaSQL` method and print the queries before running them.

**Reply by PrinceLoren** [reply]
It seems that this solution may really make sense, but  when i do this

``` javascript
await this.orm.getSchemaGenerator().updateSchema({ schema: schemaName })
```

I get the following errors like

**ERROR:  constraint "offer_user_id_foreign" of relation "offer" does not exist**

my offer.entity.ts field

``` javascript

  @ApiHideProperty()
  @Exclude({ toPlainOnly: true })
  @ManyToOne(() => User, { wrappedReference: true, nullable: true })
  user?: IdentifiedReference<User>;

```

problem is here or in settings somewhere

``` javascript
migrations: {
         disableForeignKeys: true
      },
schemaGenerator: {
        disableForeignKeys: true,
        createForeignKeyConstraints: false
 }
```

**Reply by B4nan** [reply]
FYI `disableForeignKeys` defaults to `true`, that is what adds the "session_replication_role" queries. Not sure if you actually want to disable/disallow FKs completely, as that is what `createForeignKeyConstraints` is for - it will disable generation of FKs as well as trigger removal of existing ones. I don't think you need that.

Regarding the error, I'd have to see a complete reproduction to tell you more, error message is not enough. Might be a bug caused by the `createForeignKeyConstraints` option, in such case it should be reproducible without all the "multi schema" stuff.
